### ISSUE TITLE
How to use [ResourceAuthorize("...", "...")] in Web Api?

### ISSUE THREAD

**Issue by farzad99** [IdentityServer/IdentityServer3#2326]
I modified the "MVC Authentication" sample to be able to make an ajax call from the browser directly to the Api project using access_token. So I tested the [Authorize] on the method and it works. Then I changed it to [Authorize(Roles = "Geek")] and I got a 401 response; even though, the user has a "Geek" role. So I decided to do what was done in the the MVC project using the [ResourceAuthorize("Read", "ContactDetails")] attribute. I added the same file to the Web Api project and added the following line to the Startup.cs of the Web Api project:

app.UseResourceAuthorization(new AuthorizationManager());

When I tested it I realized that filter isn't reached so the [ResourceAuthorize...] is ignored. 

So my question is: 
1. How I can get this filter [ResourceAuthorize("Read", "ContactDetails")] to work for Web Api?
2. What do I need to do to get [Authorize(Roles = "Geek")] to work?

**Comment by ptr1120** [comment]
For 1. you will have to implement a AuthorizationManager that handles the resource autorizations. [Here](https://github.com/IdentityModel/Thinktecture.IdentityModel.45/tree/master/Samples/ClaimsBasedAuthorization) is a sample (suddenly a little bit out-dated). 

For 2. A very good step-by-step guide with code sample is [Part 1 - MVC Authentication & Authorization](https://identityserver.github.io/Documentation/docsv2/overview/mvcGettingStarted.html)

**Comment by farzad99** [comment]
Thank you for your response. 

For 1. I have implemented the AuthorizationManager in the Api project of the "MVC Authentication" sample (https://identityserver.github.io/Documentation/docsv2/overview/mvcGettingStarted.html). I did it just like it is done for the MVC project in the sample and I put the following line: 

app.UseResourceAuthorization(new AuthorizationManager());

in the Startup class of the Api project of the sample solution. But the code in that file is being ignored by the Api project (My breakpoints are never hit). I also noticed that all the Web Api samples that I downloaded use the [Authorize] alone and I didn't find any of the samples using [ResourceAuthorize("...", "...")] to filter a Web Api call.

For 2. The User.Identity in the MVC project has the following properties:
AuthenticationType: "Cookies"
IsAuthenticated: true
Name: "Bob"

But the User.Identity in the Api project has:
AuthenticationType: "JWT"
IsAuthenticated: true
Name: null

I did notice that in the MVC project in the Startup class that the user claims are added to the id_token. But the claims aren't added to the access_token in the Startup class of the Api project.

So in summary, I'm sorry to say that your response didn't resolve my issues.

**Comment by leastprivilege** [comment]
wrong repo.

**Comment by Angelinsky7** [comment]
Hi, i've got the excact same problem.
the ResourceAuthorize attribute is ignored in a webapi project and the CheckAccessAsync of the implementation class of ResourceAuthorizationManager is never called.
Could someone explain use why... or what to we need to do to make it work.

**Comment by UlyssesAlves** [comment]
@farzad99 Since it has been over a year since you asked this question, you've much probably solved this issue by now. If that's the case, how did you fix it? I'm experienceing the exactly same error as you described above.

**Comment by Ben-Pattinson** [comment]
Yup, same problem here. I wonder if anyone ever figured out how to do this?
